com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -0,0 +1,328 @@
+                        return true;
+                }
+            }
+            return true;
+        }
+
+        /**
+         * Resets the file collection to empty.
+         */
+        public void reset() {
+            changedFiles = new ArrayList<>();
+            fullyQualifiedNames = new ArrayList<>();
+        }
+
+        /**
+         * Answers a collection of changed class files or <code>null</code>.
+         */
+        public List<IResource> getChangedClassFiles() {
+            return changedFiles;
+        }
+
+        /**
+         * Returns a collection of qualified type names corresponding to the changed
+         * class files.
+         *
+         * @return List
+         */
+        public List<String> getQualifiedNamesList() {
+            return fullyQualifiedNames;
+        }
+
+        /**
+         * Returns the source file associated with the given type, or <code>null</code>
+         * if no source file could be found.
+         *
+         * @param project
+         *            the java project containing the classfile
+         * @param qualifiedName
+         *            fully qualified name of the type, slash delimited
+         * @param sourceAttribute
+         *            debug source attribute, or <code>null</code> if none
+         */
+        private IResource getSourceFile(IJavaProject project, String qualifiedName, String sourceAttribute) {
+            String name = null;
+            IJavaElement element = null;
+            try {
+                if (sourceAttribute == null) {
+                    element = findElement(qualifiedName, project);
+                } else {
+                    int i = qualifiedName.lastIndexOf('/');
+                    if (i > 0) {
+                        name = qualifiedName.substring(0, i + 1);
+                        name = name + sourceAttribute;
+                    } else {
+                        name = sourceAttribute;
+                    }
+                    element = project.findElement(new Path(name));
+                }
+                if (element instanceof ICompilationUnit) {
+                    ICompilationUnit cu = (ICompilationUnit) element;
+                    return cu.getCorrespondingResource();
+                }
+            } catch (CoreException e) {
+                logger.log(Level.INFO, "Failed to get exception" + e.getMessage(), e);
+            }
+            return null;
+        }
+    }
+
+    class HotCodeReplaceEvent extends DebugEvent {
+
+        public String message;
+
+        /**
+         * Constructor.
+         */
+        public HotCodeReplaceEvent(String message) {
+            super("hotCodeReplace");
+            this.message = message;
+        }
+    }
+
+    @Override
+    public void initialize(IDebugAdapterContext context, Map<String, Object> options) {
+        // Listen to the built file events.
+        ResourcesPlugin.getWorkspace().addResourceChangeListener(this, IResourceChangeEvent.POST_BUILD);
```

Comment:
where is the logic to remove this listener?

Reply:
NO place to remove this since there is no place for removing yet.

Reply:
Please mark a TODO comment of removing this change listener